src/physics/linear_filament.rs

```diff
@@ -423,12 +569,103 @@ mod test {
     use std::f64::consts::PI;
 
     use super::*;
+    use crate::testing::*;
+
+    /// Make sure the forces have the right sign
+    /// and self-forces sum to zero within discretization error
+    #[test]
+    fn test_body_force_density() {
+        let (rtol, atol) = (5e-2, 1e-10);
+
+        let ndiscr = 100; // Discretizations of circular filament into linear filaments
+
+        // Make some circular filaments
+        let (rfil, zfil, nfil) = example_circular_filaments();
+
+        // For filament self-field, use the filament roots as the observation points
+        for i in 0..rfil.len() {
+            let (ri, zi, ni) = (rfil[i], zfil[i], nfil[i]);
+
+            let (x, y, z) = discretize_circular_filament(ri, zi, ndiscr);
+            let dl = (&diff(&x)[..], &diff(&y)[..], &diff(&z)[..]);
+            let (jxbx, jxby, jxbz) = (
+                &mut vec![0.0; ndiscr - 1],
+                &mut vec![0.0; ndiscr - 1],
+                &mut vec![0.0; ndiscr - 1],
+            );
+            body_force_density_linear_filament(
+                (&x[..ndiscr - 1], &y[..ndiscr - 1], &z[..ndiscr - 1]),
+                dl,
+                &vec![ni; x.len()][..],
+                (&x[..ndiscr - 1], &y[..ndiscr - 1], &z[..ndiscr - 1]),
+                dl,
+                (jxbx, jxby, jxbz),
+            )
+            .unwrap();
 
-    /// Div/0-resistant approximate comparison
-    fn approx(truth: f64, val: f64, rtol: f64, atol: f64) -> bool {
-        let abs_err = (val - truth).abs();
-        let lim = rtol * truth.abs() + atol;
-        abs_err < lim
+            // Make sure the totals sum to zero - a magnet can't put a net force on itself
+            let jxbx_sum: f64 = jxbx.iter().sum();
+            let jxby_sum: f64 = jxby.iter().sum();
+            let jxbz_sum: f64 = jxbz.iter().sum();
+            assert!(approx(0.0, jxbx_sum, rtol, atol));
+            assert!(approx(0.0, jxby_sum, rtol, atol));
+            assert!(approx(0.0, jxbz_sum, rtol, atol));
+
+            // Make sure jxb points outward everywhere
+            for j in 0..ndiscr - 1 {
+                let r = (x[i], y[i], z[i]);
```

Comment:
This should be indexed by `j`, the iterator over points on this circle, not `i`, the iterator over circles.
Also, `r` should be `x, y, 0` so that it is the vector from the center of the circle to a point on the circle
(rather than the vector from the origin to a point on the circle, where the origin may be above or below
the center of the circle in z).

```suggestion
                let r: (f64, f64, f64) = (x[j], y[j], 0.0);
```

After applying this suggestion, I can tighten the absolute tolerance in `approx` below from `1e-7` to `1e-8`.